### PR TITLE
Extract exception factory methods into specific exceptions

### PR DESCRIPTION
**Pull request by jwage**
|      Q       |   A
|------------- | -----------
| Type         | improvement
| BC Break     | no

#### Summary

Continuing work started in #3131

### CODE REVIEW

**Comment by jwage** [comment]
@morozov I got this rebased and cleaned up. I didn't do too much else from where the PR already was when work on it was stopped. Can we merge get this merged?

**Comment by morozov** [comment]
@jwage sure, as soon as the builds are green. BTW would it make sense if you created PRs from your personal fork in order to not have Travis run the same jobs twice?

**Comment by jwage** [comment]
The failures seem not related to any of the changes in this PR. Travis fails because of sqlsrv and php 7.4

**Comment by morozov** [comment]
The SQL Server failures will probably be fixed by the changes in https://github.com/doctrine/dbal/pull/3528. Could you try rebasing?

> I wish there was another way to make it not run twice without having to work in a fork.

What's the problem with working in a fork?

**Comment by jwage** [comment]
> What's the problem with working in a fork?

Nothing, it is just different and another "hop" in the workflow that I am not used to. If this is how everyone else is doing it, then I will adapt.

**Comment by morozov** [comment]
> If this is how everyone else is doing it […]

I believe so unless it's someone's personal project.

**Comment by jwage** [comment]
This passes now after rebasing against develop.

File: lib/Doctrine/DBAL/Connection.php [hunk 1]
```diff
@@ -912,10 +919,10 @@ public function executeQuery(string $query, array $params = [], $types = [], ?Qu
      */
     public function executeCacheQuery($query, $params, $types, QueryCacheProfile $qcp) : ResultStatement
     {
-        $resultCache = $qcp->getResultCacheDriver() ?? $this->_config->getResultCacheImpl();
+        $resultCache = $qcp->getResultCacheDriver() ?: $this->_config->getResultCacheImpl();
```
Comment (morozov) [inline lib/Doctrine/DBAL/Connection.php]: @jwage is this intentional? Most likely, it comes from an incorrect conflict resolution.

**Review by morozov — COMMENTED** [review]
Looks like there's one unintentional change. Otherwise, good to go.

File: lib/Doctrine/DBAL/Connection.php (hunk 1 above)
Comment (jwage) [reply lib/Doctrine/DBAL/Connection.php]: Yes, bad conflict resolution. I will fix, one second.

**Comment by morozov** [comment]
Thank you @Majkl578 and @jwage!

**Pull request merged**